File: modulo04-agentes-autonomos/agent-loop-framework-java/src/main/java/com/iadeva/agent/core/CircuitBreaker.java
package com.iadeva.agent.core;

import org.springframework.stereotype.Component;

import java.util.concurrent.atomic.AtomicInteger;

/**
 * Circuit breaker para evitar loops infinitos com respostas inválidas da LLM.
 * Quebra após 3 respostas inválidas consecutivas, protegendo contra alucinações
 * ou falhas persistentes no modelo.
 */
@Component
public class CircuitBreaker {

    private static final int THRESHOLD = 3;

    private final AtomicInteger invalidCount = new AtomicInteger(0);

    /**
     * Registra uma resposta inválida do LLM.
     * Incrementa o contador de falhas consecutivas.
     */
    public void recordInvalid() {
        invalidCount.incrementAndGet();
    }

    /**
     * Registra uma resposta válida do LLM.
     * Reseta o contador de falhas consecutivas.
     */
    public void recordValid() {
        invalidCount.set(0);
    }

    /**
     * Verifica se o circuit breaker deve abrir (interromper o loop).
     *
     * @return true se o número de falhas consecutivas atingiu o threshold
     */
    public boolean shouldBreak() {
        return invalidCount.get() >= THRESHOLD;
    }

    /**
     * Reseta o circuit breaker para o estado inicial.
     * Deve ser chamado no início de cada nova execução do agente.
     */
    public void reset() {
        invalidCount.set(0);
    }

    /** Retorna o número atual de respostas inválidas consecutivas */
    public int getInvalidCount() {
        return invalidCount.get();
    }
}

File: modulo04-agentes-autonomos/agent-loop-framework-go/core/circuit_breaker.go
package core

import "sync/atomic"

const circuitBreakerThreshold = 3

// CircuitBreaker evita loops infinitos com respostas inválidas do LLM —
// quebra após 3 respostas inválidas consecutivas. Equivalente a
// CircuitBreaker.java.
type CircuitBreaker struct {
	invalidCount atomic.Int32
}

// NewCircuitBreaker cria um CircuitBreaker zerado.
func NewCircuitBreaker() *CircuitBreaker {
	return &CircuitBreaker{}
}

// RecordInvalid registra uma resposta inválida do LLM.
func (c *CircuitBreaker) RecordInvalid() {
	c.invalidCount.Add(1)
}

// RecordValid registra uma resposta válida, resetando o contador.
func (c *CircuitBreaker) RecordValid() {
	c.invalidCount.Store(0)
}

// ShouldBreak retorna true se o número de falhas consecutivas atingiu o
// threshold.
func (c *CircuitBreaker) ShouldBreak() bool {
	return c.invalidCount.Load() >= circuitBreakerThreshold
}

// Reset volta o circuit breaker ao estado inicial.
func (c *CircuitBreaker) Reset() {
	c.invalidCount.Store(0)
}

// InvalidCount retorna o número atual de respostas inválidas consecutivas.
func (c *CircuitBreaker) InvalidCount() int {
	return int(c.invalidCount.Load())
}
